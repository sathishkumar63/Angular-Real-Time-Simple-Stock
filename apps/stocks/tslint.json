{
  "extends": "../../tslint.json",
  "rules": {
    "directive-selector": [true, "attribute", "stocks", "camelCase"],
    "component-selector": [true, "element", "stocks", "kebab-case"]
  },
  "linterOptions": {
    "exclude": ["!**/*"]
  }
}
