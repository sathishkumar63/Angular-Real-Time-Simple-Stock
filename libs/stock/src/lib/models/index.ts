export * from './stock.constant';
export * from './stock.interface';
export * from './stock.mock.data';
