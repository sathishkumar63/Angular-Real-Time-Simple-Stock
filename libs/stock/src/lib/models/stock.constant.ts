export const stocksSymbol: string[] = [
  'GOOG',
  'GOOGL',
  'MSFT',
  'AAPL',
  'TSLA',
  'AMZN',
  'UBER',
  'DIDI',
  'META',
  'NFLX',
  'CHIK',
  'GLD',
  'IAU',
  'BTC-USD',
  'ETH-USD',
  'DOGE-USD',
  'SHIB-USD',
  'ETH-CAD',
  'USDT-USD',
  'USDC-USD',
];
