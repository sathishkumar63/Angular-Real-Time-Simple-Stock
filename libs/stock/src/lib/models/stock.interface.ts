export interface IStock {
  name?: string;
  symbol: string;
  currency?: string;
  currentPrice: number;
  dayHighPrice?: number;
  dayLowPrice?: number;
  fiftyTwoWeekLow?: number;
  fiftyTwoWeekHigh?: number;
  regularMarketVolume?: number;
  isActive?: boolean;
}
