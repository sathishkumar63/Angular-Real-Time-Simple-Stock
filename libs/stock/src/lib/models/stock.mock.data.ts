import { StocksFeatureState, StocksState } from '../+state';
import { IStock } from './stock.interface';

export const mockStocks: IStock[] = [
  {
    name: 'Alphabet',
    symbol: 'GOOG',
    currency: 'USD',
    currentPrice: 2228.55,
    dayHighPrice: 2269.94,
    dayLowPrice: 2217.22,
    fiftyTwoWeekLow: 2044.16,
    fiftyTwoWeekHigh: 3042,
    regularMarketVolume: 1567487,
  },
  {
    name: 'Microsoft',
    symbol: 'MSFT',
    currency: 'USD',
    currentPrice: 252.99,
    dayHighPrice: 260.54,
    dayLowPrice: 252.86,
    fiftyTwoWeekLow: 246.44,
    fiftyTwoWeekHigh: 349.67,
    regularMarketVolume: 31445841,
  },
];

export const mockStocksState: StocksState = {
  data: mockStocks,
  loading: false,
  loaded: true
};

export const mockStocksFeatureState: StocksFeatureState = {
  stocks: mockStocksState
};
