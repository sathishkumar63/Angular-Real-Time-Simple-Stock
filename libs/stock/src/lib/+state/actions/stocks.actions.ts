import { Action } from '@ngrx/store';
import { IStock } from '../../models';

export enum StocksActionTypes {
  GetStocks = '[Stocks] Get',
  GetStocksSuccess = '[Stocks] Get Success',
  GetStocksFail = '[Stocks] Get Fail',
  SetStocks = '[Stocks] Set',
}

export class GetStocks implements Action {
  readonly type = StocksActionTypes.GetStocks;
}

export class GetStocksSuccess implements Action {
  readonly type = StocksActionTypes.GetStocksSuccess;
  constructor(public payload: IStock[]) {}
}

export class GetStocksFail implements Action {
  readonly type = StocksActionTypes.GetStocksFail;
  constructor(public payload: Error) {}
}

export class SetStocks implements Action {
  readonly type = StocksActionTypes.SetStocks;
  constructor(public payload: IStock) {}
}

export type StocksAction =
  | GetStocks
  | GetStocksSuccess
  | GetStocksFail
  | SetStocks;
