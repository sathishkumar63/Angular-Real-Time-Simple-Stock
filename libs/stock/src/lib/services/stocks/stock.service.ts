import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { IStock, QUOTE_URL } from '../../models';

@Injectable({
  providedIn: 'root',
})
export class StockService {
  constructor(private http: HttpClient) {}

  getStocks(symbols: string[]): Observable<any> {
    // HttpRequest has been blocked by CORS policy. So, add cors-anywhere.herokuapp.com to this link.
    const uri = `https://cors-anywhere.herokuapp.com/${QUOTE_URL}symbols=${symbols?.join()}`;
    return this.http.get(uri);
  }

  formatStocksResponse(stocks: any[]): IStock[] {
    return stocks?.map((stock) => ({
      name: stock.displayName,
      symbol: stock.symbol,
      currency: stock.currency,
      currentPrice: stock.regularMarketPrice,
      dayHighPrice: stock.regularMarketDayHigh,
      dayLowPrice: stock.regularMarketDayLow,
      fiftyTwoWeekLow: stock.fiftyTwoWeekLow,
      fiftyTwoWeekHigh: stock.fiftyTwoWeekHigh,
      regularMarketVolume: stock.regularMarketVolume,
    }));
  }
}
