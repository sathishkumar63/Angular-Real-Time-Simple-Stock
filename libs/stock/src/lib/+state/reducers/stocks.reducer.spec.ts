import { initialState, stocksReducer } from './stocks.reducer';
import * as fromStocksAction from '../actions';
import { mockStocks } from '../../models';

describe('Stocks Reducer', () => {
  describe('an unknown action', () => {
    it('should return the previous state', () => {
      // arrange
      const action = {} as any;
      // act
      const result = stocksReducer(initialState, action);
      // assert
      expect(result).toBe(initialState);
    });
  });

  describe('GetStocks', () => {
    it('should return new state with loading true', () => {
      // arrange
      const action = new fromStocksAction.GetStocks();
      // act
      const result = stocksReducer(initialState, action);
      // assert
      expect(result.loading).toBe(true);
      expect(result).toMatchSnapshot();
    });
  });

  describe('GetStocksSuccess', () => {
    it('should return new state', () => {
      // arrange
      const action = new fromStocksAction.GetStocksSuccess(mockStocks);
      // act
      const result = stocksReducer(initialState, action);
      // assert
      expect(result.data).toEqual(mockStocks);
      expect(result.loaded).toBe(true);
      expect(result.loading).toBe(false);
      expect(result).toMatchSnapshot();
    });
  });

  describe('GetStocksFail', () => {
    it('should return new state with loading and loaded false', () => {
      // arrange
      const action = new fromStocksAction.GetStocksFail(null);
      // act
      const result = stocksReducer(initialState, action);
      // assert
      expect(result.loaded).toBe(false);
      expect(result.loading).toBe(false);
      expect(result).toMatchSnapshot();
    });
  });

  describe('SetStocks', () => {
    it('should update and return the new state', () => {
      // arrange
      const action = new fromStocksAction.SetStocks(mockStocks[0]);
      // act
      const result = stocksReducer(initialState, action);
      // assert
      expect(result).toMatchSnapshot();
    });
  });
});
