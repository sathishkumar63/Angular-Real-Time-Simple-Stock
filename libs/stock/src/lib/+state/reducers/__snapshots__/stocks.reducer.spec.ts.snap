// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stocks Reducer GetStocks should return new state with loading true 1`] = `
Object {
  "data": null,
  "loaded": false,
  "loading": true,
}
`;

exports[`Stocks Reducer GetStocksFail should return new state with loading and loaded false 1`] = `
Object {
  "data": null,
  "loaded": false,
  "loading": false,
}
`;

exports[`Stocks Reducer GetStocksSuccess should return new state 1`] = `
Object {
  "data": Array [
    Object {
      "currency": "USD",
      "currentPrice": 2228.55,
      "dayHighPrice": 2269.94,
      "dayLowPrice": 2217.22,
      "fiftyTwoWeekHigh": 3042,
      "fiftyTwoWeekLow": 2044.16,
      "name": "Alphabet",
      "regularMarketVolume": 1567487,
      "symbol": "GOOG",
    },
    Object {
      "currency": "USD",
      "currentPrice": 252.99,
      "dayHighPrice": 260.54,
      "dayLowPrice": 252.86,
      "fiftyTwoWeekHigh": 349.67,
      "fiftyTwoWeekLow": 246.44,
      "name": "Microsoft",
      "regularMarketVolume": 31445841,
      "symbol": "MSFT",
    },
  ],
  "loaded": true,
  "loading": false,
}
`;

exports[`Stocks Reducer SetStocks should update and return the new state 1`] = `
Object {
  "data": undefined,
  "loaded": false,
  "loading": false,
}
`;
